### PR TITLE
version 2 with extension memos and higher reliance on extensions

### PR DESCRIPTION
**Pull request by michaelficarra**
Also remove built-in optimisations. Implementations will provide their own standard optimisations. I'll update the comparison table once I pull out the optimisations into standard extensions in superpack-js.

### CODE REVIEW

**Comment by LewisJEllis** [comment]
I've taken a skim-level look at this and it seems interesting and I like the idea behind it. If you're not in a hurry to merge this, I would do an in-depth review in the next few days. I'm particularly interested in seeing the JS implementation for concrete realizations of the ideas specified here.

**Comment by michaelficarra** [comment]
@LewisJEllis See https://github.com/shapesecurity/superpack-js/pull/48 for the extension memos implementation in JS. Let me know if you have any additional questions. The basic idea here is that we've provided a generalisation that encompasses our existing built-in optimisations and allows for domain-specific optimisations and more powerful extensions to be implemented by the user. So we've removed the built-in optimisations, opting instead for them to be provided by the transcoder as optional extensions. We've also dedicated more type tag space to extensions so we don't actually lose anything efficiency-wise.

File: README.md [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
-etc.
+SuperPack is a data serialisation format. Due to its design and the
+optimisations enabled by its [extension mechanism](#extensions), it is
+particularly suitable for encoding values with repeated structure: arrays of
+booleans, strings used more than once, objects which have the same shape, etc.
 
 A comparison of schemaless serialisation formats run on 1000 [public domain book records](https://www.npmjs.com/package/public-domain-nypl-captures)
 from the [New York Public Library Digital Collections API](http://api.repo.nypl.org/):
 
+**TODO: split into SuperPack with no optimisations and SuperPack with only built-in optimisations**
```
Comment (disnet) [inline README.md]: doing this TODO before or after you merge this PR?

File: README.md [hunk 2]
```diff
@@ -444,68 +417,60 @@ transcoder.extend(
...
-| 2kB        | 500kB         |
-| 2.83kB     | 1mB           |
-| 4kB        | 2mB           |
-| 40kB       | 200mB         |
-| 400kB      | 20gB          |
+Memos are only prepended for extensions that define the optional `memo`
+function. Memos are encoded and prepended to the SuperPack payload in
+descending extension point order. During the encoding of a memo, extensions
+that define the optional `memo` function must not be applied if their extension
+point is greater than that of the extension whose memo is being encoded.
```
Comment (disnet) [inline README.md]: this section seems a bit underspecified. e.g. what is a "memo" exactly? What does it mean to "encode and prepend" a memo?

File: README.md (hunk 1 above)
Comment (michaelficarra) [reply README.md]: Before. After I implement in superpack-js.

File: README.md [hunk 3]
```diff
@@ -408,17 +408,24 @@ let transcoder = new SuperPackTranscoder;
...
 
-A SuperPack encoder must not recursively apply an extension when encoding the
-result of its own serialisation.
+If the extension uses state to keep track of values it has marked as
+candidates, it can use that state in conjunction with the optional
+`shouldSerialise` method to determine if an extension should be applied to the
+value. `shouldSerialise` will only ever be called once all candidates have been
+determined (see Extensions and Recursion for an exception to this). If
+`shouldSerialise` is not provided, the extension behaves as if it was given a
+function which always returns `true`.
```
Comment (disnet) [inline README.md]: This is an API concern not a spec concern right?

File: README.md [hunk 4]
```diff
@@ -427,30 +434,31 @@ values encoded by an extension. This is mostly useful for deduplication and
...
+    return this.symbols[n];
+  }
+  memo() { return this.symbols.map(sym => String(sym).slice(7, -1)); }
+}
 
-let encoded = encode(data, { extensions: { 0: symbolExtension } });
+let encoded = encode(data, { extensions: { 0: SymbolExtension } });
 ```
 
 Memos are only prepended for extensions that define the optional `memo`
```
Comment (disnet) [inline README.md]: > Memos are only prepended for extensions that define the optional `memo` function. Memos are encoded and prepended to the SuperPack payload in descending extension point order.

This is an odd mixing of specification requirements ("descending extension point order") with an implementation specific API suggestion ("optional `memo` function").

File: README.md (hunk 3 above)
Comment (michaelficarra) [reply README.md]: You're right.

File: README.md (hunk 4 above)
Comment (michaelficarra) [reply README.md]: You're right, I was in two places at once when I wrote this.

**Review by disnet — APPROVED** [review]

**Pull request merged**